File: tinyredis-core/src/main/java/com/github/cjqcn/tinyredis/core/command/impl/GetCommand.java
package com.github.cjqcn.tinyredis.core.command.impl;

import com.github.cjqcn.tinyredis.core.client.RedisClient;
import com.github.cjqcn.tinyredis.core.command.RedisCommand;
import com.github.cjqcn.tinyredis.core.db.RedisDb;
import com.github.cjqcn.tinyredis.core.exception.RedisException;
import com.github.cjqcn.tinyredis.core.struct.RedisObject;
import com.github.cjqcn.tinyredis.core.struct.impl.StringRedisObject;
import com.github.cjqcn.tinyredis.core.util.DBUtil;

public class GetCommand extends AbstractCommand implements RedisCommand {

    private final String key;

    public GetCommand(RedisClient redisClient, String key) {
        super(redisClient);
        this.key = key;
    }

    @Override
    public void execute() {
        execute0(redisClient, key);
    }

    public void execute0(RedisClient redisClient, String key) {
        RedisDb db = redisClient.curDb();
        RedisObject value = DBUtil.getValueWithExpire(db, key);
        if (value == null) {
            redisClient.stream().response(SimpleStringResponse.GET_NULL);
            return;
        }
        if (value instanceof StringRedisObject) {
            redisClient.stream().responseString(value.get().toString());
        } else {
            throw RedisException.WRONG_TYPE_OPERATION;
        }
    }

    public static GetCommand build(RedisClient redisClient, String value) {
        // TODO cache
        return new GetCommand(redisClient, value);
    }

    @Override
    public String toString() {
        return "GetCommand{" + key + '}';
    }
}

File: tinyredis-core/src/main/java/com/github/cjqcn/tinyredis/core/command/impl/GetCommand.kt
package com.github.cjqcn.tinyredis.core.command.impl

import com.github.cjqcn.tinyredis.core.client.RedisClient
import com.github.cjqcn.tinyredis.core.command.RedisCommand
import com.github.cjqcn.tinyredis.core.command.RedisResponse
import com.github.cjqcn.tinyredis.core.exception.RedisException
import com.github.cjqcn.tinyredis.core.struct.RedisObject
import com.github.cjqcn.tinyredis.core.struct.StringObject
import com.github.cjqcn.tinyredis.core.struct.TtlObject

class GetCommand(private val redisClient: RedisClient, private val key: RedisObject) : RedisCommand {

    override fun client(): RedisClient {
        return redisClient;
    }

    override fun execute(): RedisResponse {
        val db = client().db()
        val ttl = db.expires().get(key)
        if (ttl is TtlObject && !ttl.isLive()) {
            db.expires().del(key)
            db.dict().del(key)
        }
        val value = db.dict().get(key) ?: return GetResponse(StringObject.NULL)
        if (value !is StringObject) {
            throw RedisException.WRONG_TYPE_OPERATION
        }
        return GetResponse(value)
    }

    override fun name(): String {
        return "get";
    }
}